chainer/reporter.py

```diff
@@ -0,0 +1,327 @@
+    to report the observed values (e.g., training loss, accuracy, activation
+    statistics, etc.).
+
+    Attributes:
+        observation: Dictionary of observed values.
+
+    """
+    def __init__(self):
+        self._observer_names = {}
+        self.observation = {}
+
+    def __enter__(self):
+        """Makes this reporter object current."""
+        _reporters.append(self)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Recovers the previous reporter object to the current."""
+        _reporters.pop()
+
+    @contextlib.contextmanager
+    def scope(self, observation):
+        """Creates a scope to report observed values to the given dictionary.
+
+        This is a context manager to be passed to ``with`` statements. In this
+        scope, the observation dictionary is changed to the given one.
+
+        It also makes this reporter object current.
+
+        Args:
+            observation (dict): Observation dictionary. All observations
+                reported inside of the ``with`` statement are written to this
+                dictionary.
+
+        """
+        old = self.observation
+        self.observation = observation
+        self.__enter__()
+        yield
+        self.__exit__(None, None, None)
+        self.observation = old
+
+    def add_observer(self, name, observer):
+        """Registers an observer of values.
+
+        Observer defines a scope of names for observed values. Values observed
+        with the observer are registered with names prefixed by the observer
+        name.
+
+        Args:
+            name (str): Name of the observer.
+            observer: The observer object. Note that the reporter distinguishes
+                the observers by their object ids (i.e., ``id(owner)``), rather
+                than the object equality.
+
+        """
+        self._observer_names[id(observer)] = name
+
+    def add_observers(self, prefix, observers):
+        """Registers multiple observers at once.
+
+        This is a convenient method to register multiple objects at once.
+
+        Args:
+            prefix (str): Prefix of each name of observers.
+            observers: Iterator of name and observer pairs.
+
+        """
+        for name, observer in observers:
+            self._observer_names[id(observer)] = prefix + name
+
+    def report(self, values, observer=None):
+        """Reports observed values.
+
+        The values are written with the key, prefixed by the name of the
+        observer object if given.
+
+        Args:
+            values (dict): Dictionary of observed values.
+            observer: Observer object. Its object ID is used to retrieve the
+                observer name, which is used as the prefix of the registration
+                name of the observed value.
+
+        """
+        if observer is not None:
+            observer_name = self._observer_names[id(observer)]
```

Comment:
@beam2d I tried implementing the data parallel mnist example using the `trainer2 `branch. I would get the following error:

> Traceback (most recent call last):
>   File "train_mnist_parallel2.py", line 215, in <module>
>     main()
>   File "train_mnist_parallel2.py", line 211, in main
>     trainer.run()
>   File "/u/amitaid/code/chainer/chainer/training/trainer.py", line 244, in run
>     update()
>   File "/u/amitaid/code/chainer/chainer/training/updater.py", line 162, in update
>     self._update_func(self)
>   File "train_mnist_parallel2.py", line 126, in update
>     losses.append(model(*in_vars))
>   File "/u/amitaid/code/chainer/chainer/links/model/classifier.py", line 57, in __call__
>     reporter.report({'loss': self.loss}, self)
>   File "/u/amitaid/code/chainer/chainer/reporter.py", line 197, in report
>     current.report(values, observer)
>   File "/u/amitaid/code/chainer/chainer/reporter.py", line 132, in report
>     observer_name = self._observer_names[id(observer)]
> KeyError: 47693283314192

This is caused by reports sent from the classifier on the second GPU. I guess it was not added by the `Evaluator`. My solution was to first check if the `observer` exists:
`if id(observer) in self._observer_names.keys():`